src/polycubed/src/server/Resources/Endpoint/Hateoas.h

```diff
@@ -0,0 +1,201 @@
+
+    /**
+     * using a key, that member function will search the right value in the
+     * response that are provided from the cube in order to resolve
+     * dynamic endpoints.
+     *
+     * @param[in] errors        the response that will be sent to the client
+     * @param[out] js_hateoas   the json where Hateoas records will be stored
+     * @param[in] base          current level endpoint
+     * @param[in] key           the key that will be used to find next level values
+     *                          that will make up final endpoints
+     * @returns                 true if the key was found, false otherwise
+     */
+    static bool writeNextlevel(std::vector<Response> &errors, json &js_hateoas,
+            std::string& base, std::string &key);
+
+    /**
+     * provides the token of the specified level for a given endpoint
+     *
+     * @param[in] endpoint      the endpoint to process
+     * @param[in] level         the level required
+     * @returns                 the token of the required level for a given endpoint
+     */
+    static std::string getToken_level(const std::string &endpoint,
+            const size_t level);
+
+    /**
+     * check if the token provided is a key or not
+     *
+     * @param[in] token             the token to process
+     * @returns                 true if the token is a key, false otherwise
+     */
+    static bool is_key(std::string &token);
+
+    /**
+     * split the href provided to tokens
+     *
+     * @param[in] href              the href to split
+     * @returns                 tokens
+     */
+    static std::vector<std::string> split_to_tokens(const std::string &href);
+
+    /// map containing the routes only of the level just after the root
+    static std::unordered_map<std::string, std::string> root_routes;
+
+
+public:
+
+    Hateoas(Hateoas const&)         = delete;
+    void operator=(Hateoas const&)  = delete;
+
+    /**
+     * Hateoas support for a ParentResource.
+     * Children's links are attached to the response.
+     *
+     * @param[in] request       the request of the client
+     * @param[in,out] errors    the response to which the links will be added
+     * @param[in] host          server host address
+     * @param[in] port          server port
+     * @param[in] children      children of the processed ParentResource node
+     */
+    static void HateoasSupport(const Request &request,
+            std::vector<Response> &errors,
+            const std::string &host, const std::string &port,
+            const std::vector<std::shared_ptr<polycube::polycubed
+            ::Rest::Resources::Body::Resource>> &children);
+
+    /**
+     * Hateoas support for a ListResource.
+     * Dynamic endpoint routes are resolved by processing
+     * the response json of the cube list resource.
+     *
+     * @param[in] request       the request of the client
+     * @param[in,out] errors    the response to which the links will be added
```

Comment:
Why a "response" is called "error"? Is there any better name?

Reply:
Actually in the polycube code the response is called "error" or "errors".
Here some examples:

- https://github.com/polycube-network/polycube/blob/df3c16e0907b83b6852f62e006c2da94a82618d2/src/polycubed/src/server/Resources/Endpoint/ParentResource.cpp#L173
- https://github.com/polycube-network/polycube/blob/df3c16e0907b83b6852f62e006c2da94a82618d2/src/polycubed/src/server/Resources/Endpoint/ListResource.cpp#L197

 and the ResponseWriter object is called "response".

I used the same convention in Hateoas but I can change it and call the response as "response"

Reply:
Davide, what about cleaning up this in Polycube, so that it becomes easier to understand?
However, if you agree I would close this PR as is (leaving this as "error"), then open a new PR (polycube-wide) with this improvement.

Reply:
Well, no problems for me about that